File: Flink0105/src/main/scala/com/atguigu/day5/ListStateExample.scala
package com.atguigu.day5

import com.atguigu.day2.{SensorReading, SensorSource}
import org.apache.flink.api.common.state.{ListStateDescriptor, ValueStateDescriptor}
import org.apache.flink.api.scala.typeutils.Types
import org.apache.flink.runtime.state.StateBackend
import org.apache.flink.runtime.state.filesystem.FsStateBackend
import org.apache.flink.streaming.api.functions.KeyedProcessFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.util.Collector

import scala.collection.mutable.ListBuffer

object ListStateExample {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)

    // 每个10s做一次检查点操作
    env.enableCheckpointing(10000L)
    // windows: file:\\ + 绝对路径
    env.setStateBackend(new FsStateBackend("file:///Users/yuanzuo/Desktop/flink-tutorial/Flink0105/src/main/resources/checkpoint"))

    // env.setStateBackend(new FsStateBackend("file:///").asInstanceOf[StateBackend])

    val stream = env
      .addSource(new SensorSource)
      .filter(_.id.equals("sensor_1"))
      .keyBy(_.id)
      .process(new Keyed)

    stream.print()

    env.execute()
  }

  class Keyed extends KeyedProcessFunction[String, SensorReading, String] {
    lazy val listState = getRuntimeContext.getListState(
      new ListStateDescriptor[SensorReading]("list-state", Types.of[SensorReading])
    )

    lazy val timer = getRuntimeContext.getState(
      new ValueStateDescriptor[Long]("timer", Types.of[Long])
    )

    override def processElement(value: SensorReading, ctx: KeyedProcessFunction[String, SensorReading, String]#Context, out: Collector[String]): Unit = {
      listState.add(value) // 将value添加到列表状态中
      if (timer.value() == 0L) {
        val ts = ctx.timerService().currentProcessingTime() + 10 * 1000L
        ctx.timerService().registerProcessingTimeTimer(ts)
        timer.update(ts)
      }
    }

    override def onTimer(timestamp: Long, ctx: KeyedProcessFunction[String, SensorReading, String]#OnTimerContext, out: Collector[String]): Unit = {
      // 不能直接对列表状态变量进行计数
      val readings: ListBuffer[SensorReading] = new ListBuffer()
      // 隐式类型转换必须导入
      import scala.collection.JavaConversions._
      for (r <- listState.get) {
        readings += r
      }
      out.collect("当前时刻列表状态变量里面共有 " + readings.size + "条数据！")
      timer.clear()
    }
  }
}

File: flink-code-java-0421/src/main/java/com/atguigu/day05/ListStateExample.java
package com.atguigu.day05;

import com.atguigu.day02.util.SensorReading;
import com.atguigu.day02.util.SensorSource;
import org.apache.flink.api.common.state.ListState;
import org.apache.flink.api.common.state.ListStateDescriptor;
import org.apache.flink.api.common.state.ValueState;
import org.apache.flink.api.common.state.ValueStateDescriptor;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
import org.apache.flink.util.Collector;

public class ListStateExample {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        DataStreamSource<SensorReading> stream = env.addSource(new SensorSource());

        stream
                .filter(r -> r.id.equals("sensor_1"))
                .keyBy(r -> r.id)
                .process(new KeyedProcessFunction<String, SensorReading, String>() {
                    private ValueState<Long> timer;
                    private ListState<SensorReading> readingList;

                    @Override
                    public void open(Configuration parameters) throws Exception {
                        super.open(parameters);
                        timer = getRuntimeContext().getState(
                                new ValueStateDescriptor<Long>("timer", Types.LONG)
                        );
                        readingList = getRuntimeContext().getListState(
                                new ListStateDescriptor<SensorReading>("reading-list", SensorReading.class)
                        );
                    }

                    @Override
                    public void processElement(SensorReading sensorReading, Context context, Collector<String> collector) throws Exception {
                        readingList.add(sensorReading);
                        if (timer.value() == null) {
                            context.timerService().registerProcessingTimeTimer(context.timerService().currentProcessingTime() + 10 * 1000L);
                            timer.update(context.timerService().currentProcessingTime() + 10 * 1000L);
                        }
                    }

                    @Override
                    public void onTimer(long timestamp, OnTimerContext ctx, Collector<String> out) throws Exception {
                        super.onTimer(timestamp, ctx, out);
                        long count = 0L;
                        for (SensorReading r : readingList.get()) {
                            count += 1;
                        }
                        timer.clear();
                        out.collect("当前流中有 " + count + " 条元素");
                    }
                })
                .print();

        env.execute();
    }
}
